set nonzero exit status when running script fails

Fairly opened the pull request:
closes #6912

Fairly committed:
Merge pull request #1 from ipython/master

a

Fairly committed:
Fix #6912 "IPython not setting exit status if script errors".
Only handle the issue related to .py scripts.

1. interactiveshell: The method safe_execfile is not totally safe now. Errors of user scripts  are raised to a higher level, where we can decide what to do.

2. execution: The %run magic catches exceptions from safe_execfile.

3. shellapp: The _run_cmd_line_code method catches exceptions from safe_execfile and sets the right exit status.

Fairly committed:
Merge pull request #2 from ipython/master

b

minrk commented:
I don't think there are any changes needed outside `shellapp`. I think we just need to be passing `raise_exceptions=True` when we call safe_execfile. Can you give that a try?

Fairly commented:
Yes, I think you are right. It didn't occur to me. But I found another problem in safe_execfile. When I passed `raise_exceptions=True` to it, it would raise an `SystemExit` exception even the user script returned normally with a `sys.exit(0)` call. Anyway, I will fix it in the next commit.

Fairly committed:
1. Change execution.py back.

2. Revert some changes in safe_execfile, then modify the excetion
handling logic.

minrk commented on IPython/core/interactiveshell.py:
```diff
@@ -2674,10 +2674,11 @@ def safe_execfile(self, fname, *where, **kw):
                 # For other exit status, we show the exception unless
                 # explicitly silenced, but only in short form.
-                if kw['raise_exceptions']:
-                    raise
```
This now ignores `SystemExit(0)`. Is that a deliberate change?

Fairly replied:
Yeah. I set `raise_exceptions` to `True` in `shellapp`. If `safe_execfile` doesn't ignore `SystemExit(0)`, an `SystemError` exception will be raised even user's scripts exit normally. 
Actually, I think this is `safe_execfile` intended to do, as it says in comments.
```python
# If the call was made with 0 or None exit status (sys.exit(0)
# or sys.exit() ), don't bother showing a traceback, as both of
# these are considered normal by the OS:
# > python -c'import sys;sys.exit(0)'; echo $?
# 0
# > python -c'import sys;sys.exit()'; echo $?
# 0
# For other exit status, we show the exception unless
# explicitly silenced, but only in short form.
```

minrk replied:
Terrific, thanks for clarifying.

minrk commented:
Thanks, merged as a4d98464c801c7475c8b2169e9a372970b72ebfb to avoid the extra merge commits.

fHachenberg commented:
So how can one interrupt execution of an IPython script and return a statuscode != 0 ?

raise SystemExit(4) 

does not stop execution, because SystemExit is caught unconditionally and never reraised in InteractiveShell.run_code.
